compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConvertToDma.cpp

```diff
@@ -146,129 +157,452 @@ LogicalResult unPackDmaInputs(IREE::LinalgExt::UnPackOp unPackOp,
+      // Case 3.
+      if (auto lcAttr = dyn_cast<Attribute>(lc)) {
+        int64_t c = cast<IntegerAttr>(lcAttr).getInt();
+        if (c != 0) {
+          lc = builder.create<arith::AddIOp>(loc, getConstant(c), term)
+                   .getResult();
+        } else {
+          lc = term;
+        }
+      }
+      // Case 4.
+      else {
+        lc = builder.create<arith::AddIOp>(loc, cast<Value>(lc), term)
+                 .getResult();
+      }
     }
-    offsets = subviewOp.getMixedOffsets();
-    auto [stridesI64, baseOffset] =
-        getStridesAndOffset(subviewOp.getSource().getType());
-    if (baseOffset != 0) {
-      auto message = llvm::formatv(
-          "has non-zero base offset {0} that is not supported by the "
-          "current pass: requires testing and possible code changes.",
-          baseOffset);
-      return subviewOp->emitOpError(message);
+  }
+  return lc;
+}
+
+/// Update the offsets, sizes, and strides from a collapse shape operation.
+LogicalResult updateFromCollapseShape(memref::CollapseShapeOp collapseOp,
+                                      SmallVector<OpFoldResult> &offsets,
+                                      SmallVector<OpFoldResult> &sizes,
+                                      SmallVector<OpFoldResult> &strides) {
+  auto reassociationIndices = collapseOp.getReassociationIndices();
+  ArrayRef<int64_t> inputShape = collapseOp.getSrcType().getShape();
+  ArrayRef<int64_t> resultShape = collapseOp.getType().getShape();
+  uint64_t resultRank = resultShape.size();
+  MLIRContext *ctx = collapseOp.getContext();
+
+  // Set strides to inner-most stride in each reassocation group.
+  //
+  // Example: Consider a 2x3x5x7 tensor, with strides [70,35,7,1]. If this 
+  // is collapsed to shape 6x35, the srides are [35,1]. The reassociation
+  // groups are [0,1] and [2,3], and so we've just taken the inner-most
+  // strides in each group.
+  for (auto reassociation : llvm::enumerate(reassociationIndices)) {
+    uint64_t index = reassociation.index();
+    uint64_t dim = reassociation.value().back();
+    strides[index] = strides[dim];
+  }
+  strides.resize(resultRank);
+
+  // Set sizes to output shape, and check that all dims are static.
+  sizes.clear();
+  for (int64_t dim : resultShape) {
+    if (dim == ShapedType::kDynamic) {
+      return collapseOp.emitOpError(
+          "has a dynamic shape which is currently unsupported.");
     }
-    strides = getAsIndexOpFoldResult(ctx, stridesI64);
-    operandOp = subviewOp.getSource().getDefiningOp();
-    sizes = subviewOp.getMixedSizes();
-    if (llvm::any_of(sizes, [](OpFoldResult fr) {
-          return !getConstantIntValue(fr).has_value();
-        })) {
-      return subviewOp->emitOpError(
-          " has dynamic shape that is not supported by the target dma op.");
+    sizes.push_back(getAsIndexOpFoldResult(ctx, dim));
+  }
+
+  // Offsets - merge reassocation groups by taking linear combinations of the
```

Comment:
I don't trust this. 

Consider 1x5 with stride = [10,1] offset = [100,0] being collapsed to 5. New offset should be 10*100, not 5*100.